GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs

```diff
@@ -353,42 +353,117 @@ public void SupersededFileAddedToModifiedPathsAndSkipWorktreeBitCleared()
+            fileOutsideRepo.ShouldNotExistOnDisk(this.fileSystem);
             this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
-        }
-
+        }
+
         [TestCase, Order(17)]
         public void FileMovedFromInsideRepoToOutside()
         {
             string fileName = "InsideRepoToOutside.txt";
-            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
-            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
-            fileInsideRepo.ShouldBeAFile(this.fileSystem);
+            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
+            fileInsideRepo.ShouldBeAFile(this.fileSystem);
             this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
-
-            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
-            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
+
+            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
             fileInsideRepo.ShouldNotExistOnDisk(this.fileSystem);
-            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
-            this.Enlistment.WaitForBackgroundOperations();
+            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+        }
+
+        [TestCase, Order(18)]
+        public void HardlinkFromOutsideRepoToInside()
+        {
+            string fileName = "OutsideRepoToInside_FileForHardlink.txt";
+            string fileOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.WriteAllText(fileOutsideRepo, "Contents for the new file");
+            fileOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileNameLink = "OutsideRepoToInside_RepoLink.txt";
+            string fileLinkInsideRepo = this.Enlistment.GetVirtualPathTo(fileNameLink);
+            this.fileSystem.CreateHardLink(fileLinkInsideRepo, fileOutsideRepo);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileNameLink);
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+            fileNameLink.ShouldBeAFile(this.fileSystem);

```

Comment:
It looks like there is a copy/paste error on this line that is causing your functional test failure:

`fileNameLink.ShouldBeAFile` should be `fileLinkInsideRepo.ShouldBeAFile`